Framework/DataHandling/src/LoadInstrument.cpp

```diff
@@ -95,74 +97,103 @@ void LoadInstrument::init() {
     if (xml) {
-      parser = InstrumentDefinitionParser(m_filename, m_instName, *xml);
+      parser = InstrumentDefinitionParser(filename, instname, *xml);
     } else {
       throw std::invalid_argument("The instrument XML passed cannot be "
                                   "casted to a standard string.");
     }
-  }
-  // otherwise we need either Filename or InstrumentName to be set
-  else {
-    // Retrieve the filename from the properties
-    if (m_filename.empty()) {
+    idfInstrumentLoader(ws, filename, parser);
+
+  } else {
+    // This part of the loader searches through the instrument directories for
+    // a valid IDF or Nexus geometry file
+
+    // The first step is to define a valid filename
+
+    // If the filename is empty, try to find a file from the instrument name
+    if (filename.empty()) {
       // look to see if an Instrument name provided in which case create
-      // IDF filename on the fly
-      if (m_instName.empty()) {
+      // filename on the fly
+      if (instname.empty()) {
         g_log.error("Either the InstrumentName or Filename property of "
                     "LoadInstrument most be specified");
         throw Kernel::Exception::FileError(
             "Either the InstrumentName or Filename property of LoadInstrument "
-            "most be specified to load an IDF",
-            m_filename);
+            "must be specified to load an instrument",
+            filename);
       } else {
-        const std::string date = m_workspace->getWorkspaceStartDate();
-        m_filename = ExperimentInfo::getInstrumentFilename(m_instName, date);
+        filename = ExperimentInfo::getInstrumentFilename(
+            instname, ws->getWorkspaceStartDate());
+        setPropertyValue("Filename", filename);
       }
     }
-
-    if (m_filename.empty()) {
+    if (filename.empty()) {
       throw Exception::NotFoundError(
-          "Unable to find an Instrument Definition File for", m_instName);
+          "Unable to find an Instrument File for instrument: ", instname);
     }
 
-    // Remove the path from the filename for use with the InstrumentDataService
-    const std::string::size_type stripPath = m_filename.find_last_of("\\/");
-    std::string instrumentFile =
-        m_filename.substr(stripPath + 1, m_filename.size());
-    // Strip off "_Definition.xml"
-    m_instName = instrumentFile.substr(0, instrumentFile.find("_Def"));
-
-    // Initialize the parser with the the XML text loaded from the IDF file
-    parser = InstrumentDefinitionParser(m_filename, m_instName,
-                                        Strings::loadFile(m_filename));
+    // Now that we have a file name, decide whether to use Nexus or IDF loading
+    if (LoadGeometry::isIDF(filename)) {
+      // Remove the path from the filename for use with the
+      // InstrumentDataService
+      const std::string::size_type stripPath = filename.find_last_of("\\/");
```

Comment:
This stuff isn't ideal (getting name from instrument file) but you didn't write it so lets leave it as is.